File: packages/go/structural.go
package gof

import (
	"iter"
	"sync"
)

// Adapt is Adapter: expose an incompatible value through the interface you want.
//
// Thin on purpose. In Go the idiomatic adapter is a small struct that implements
// your interface, or interface embedding, and you should usually write that. This
// names the intent where a func literal is enough.
func Adapt[Source, API any](source Source, build func(Source) API) API {
	return build(source)
}

// Decorate is Decorator: wrap a func in layers — retry, cache, log — without
// touching it. The first wrapper is the outermost, so the arguments read in the
// order they run.
//
//	charge := gof.Decorate(rawCharge, withLog, withRetry(3))
func Decorate[In, Out any](
	fn func(In) Out,
	wrappers ...func(func(In) Out) func(In) Out,
) func(In) Out {
	wrapped := fn
	for i := len(wrappers) - 1; i >= 0; i-- {
		wrapped = wrappers[i](wrapped)
	}
	return wrapped
}

// Composite lets a tree of nodes be treated like a single node.
type Composite[T any] struct {
	Value    T
	children []*Composite[T]
}

// NewComposite returns a node holding value, with optional children.
func NewComposite[T any](value T, children ...*Composite[T]) *Composite[T] {
	return &Composite[T]{Value: value, children: children}
}

// Add appends children and returns the node, so calls chain.
func (c *Composite[T]) Add(nodes ...*Composite[T]) *Composite[T] {
	c.children = append(c.children, nodes...)
	return c
}

// Remove drops one child, reporting whether it was there.
func (c *Composite[T]) Remove(child *Composite[T]) bool {
	for i, existing := range c.children {
		if existing == child {
			c.children = append(c.children[:i], c.children[i+1:]...)
			return true
		}
	}
	return false
}

// Children returns the direct children.
func (c *Composite[T]) Children() []*Composite[T] { return c.children }

// Walk yields the subtree depth-first, self first. It is an [iter.Seq], so a
// `break` in the range loop stops the traversal instead of walking the rest.
func (c *Composite[T]) Walk() iter.Seq[*Composite[T]] {
	return func(yield func(*Composite[T]) bool) {
		c.walk(yield)
	}
}

func (c *Composite[T]) walk(yield func(*Composite[T]) bool) bool {
	if !yield(c) {
		return false
	}
	for _, child := range c.children {
		if !child.walk(yield) {
			return false
		}
	}
	return true
}

// Sum adds of(node.Value) over the whole subtree.
func (c *Composite[T]) Sum(of func(T) float64) float64 {
	total := 0.0
	for node := range c.Walk() {
		total += of(node.Value)
	}
	return total
}

// Size counts the nodes in the subtree, including this one.
func (c *Composite[T]) Size() int {
	count := 0
	for range c.Walk() {
		count++
	}
	return count
}

// Flyweight shares one value per key instead of re-creating equal values.
//
//	types := gof.NewFlyweight(func(name string) *TreeType { return &TreeType{Name: name} })
//	types.Get("oak") == types.Get("oak")   // same pointer
type Flyweight[K comparable, V any] struct {
	factory func(K) V
	mu      sync.Mutex
	cache   map[K]V
}

// NewFlyweight returns a Flyweight that builds values with factory.
func NewFlyweight[K comparable, V any](factory func(K) V) *Flyweight[K, V] {
	return &Flyweight[K, V]{factory: factory, cache: make(map[K]V)}
}

// Get returns the shared value for key, building it on first use. It is safe to
// call from several goroutines, and factory runs once per key.
func (f *Flyweight[K, V]) Get(key K) V {
	f.mu.Lock()
	defer f.mu.Unlock()
	value, ok := f.cache[key]
	if !ok {
		value = f.factory(key)
		f.cache[key] = value
	}
	return value
}

// Len reports how many distinct keys are held.
func (f *Flyweight[K, V]) Len() int {
	f.mu.Lock()
	defer f.mu.Unlock()
	return len(f.cache)
}

// Clear drops every cached value.
func (f *Flyweight[K, V]) Clear() {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.cache = make(map[K]V)
}

// Bridge is a stable abstraction whose implementation can be swapped underneath.
//
// Callers hold the Bridge and read [Bridge.API] per call, so a [Bridge.Swap]
// reaches everyone who already has it. Passing the implementation into a
// constructor — the textbook version — makes every holder re-wire instead.
type Bridge[Impl, API any] struct {
	build func(Impl) API
	mu    sync.RWMutex
	api   API
}

// NewBridge returns a Bridge over the given implementation.
func NewBridge[Impl, API any](build func(Impl) API, implementation Impl) *Bridge[Impl, API] {
	return &Bridge[Impl, API]{build: build, api: build(implementation)}
}

// API returns the current abstraction. Read it per call, not once into a variable,
// or a later Swap will not reach you.
func (b *Bridge[Impl, API]) API() API {
	b.mu.RLock()
	defer b.mu.RUnlock()
	return b.api
}

// Swap replaces the implementation behind the abstraction.
func (b *Bridge[Impl, API]) Swap(implementation Impl) {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.api = b.build(implementation)
}

// Facade and Proxy have no helper in Go, because the standard library already does
// both properly with sync.OnceValue — a struct whose subsystems are built on first
// use:
//
//	type Checkout struct {
//		payments func() *PaymentClient // sync.OnceValue(func() *PaymentClient { ... })
//		mail     func() *Mailer
//	}
//
//	func (c *Checkout) Pay(cents int) string { return c.payments().Charge(cents) }
//
// OnceValue is lazy, runs the factory once and is goroutine-safe. Nothing a helper
// could add. See the README table.

File: packages/rust/src/structural.rs
//! Structural patterns.
//!
//! Three ship no helper, because the language already does them:
//!
//! - **Adapter** — a newtype implementing your trait, or trait delegation. Four
//!   lines, checked by the compiler.
//! - **Facade** — a struct whose subsystems are [`std::sync::OnceLock`] fields,
//!   each built on first use.
//! - **Proxy** — [`std::sync::LazyLock`], which is lazy, runs once and is `Sync`.

use std::collections::HashMap;
use std::hash::Hash;
use std::sync::{Arc, Mutex, RwLock};

/// A boxed function of one argument: the unit [`decorate`] composes.
pub type Layer<In, Out> = Box<dyn Fn(In) -> Out>;

/// One layer of a [`decorate`] stack: it receives the function underneath it and
/// returns the function that replaces it.
pub type Wrapper<In, Out> = Box<dyn Fn(Layer<In, Out>) -> Layer<In, Out>>;

/// Decorator — wrap a function in layers (retry, cache, log) without touching it.
/// The first wrapper is the outermost, so the list reads in the order it runs.
///
/// ```
/// use gof_patterns::{decorate, Layer};
///
/// use gof_patterns::Wrapper;
///
/// let log: Wrapper<u32, String> =
///     Box::new(|next| Box::new(move |n| format!("log({})", next(n))));
/// let twice: Wrapper<u32, String> = Box::new(|next| Box::new(move |n| next(n * 2)));
///
/// let call = decorate(Box::new(|n: u32| n.to_string()), vec![log, twice]);
/// assert_eq!(call(3), "log(6)");
/// ```
pub fn decorate<In, Out>(
    function: Layer<In, Out>,
    wrappers: Vec<Wrapper<In, Out>>,
) -> Layer<In, Out> {
    wrappers
        .into_iter()
        .rev()
        .fold(function, |next, wrap| wrap(next))
}

/// Composite — treat a tree of nodes like a single node.
///
/// [`Composite::walk`] is a lazy iterator over an explicit stack, so `find` and
/// `take` stop the traversal instead of visiting the whole tree.
///
/// ```
/// use gof_patterns::Composite;
///
/// let mut root = Composite::new(10.0);
/// root.add(Composite::new(5.0));
/// assert_eq!(root.sum(|price| *price), 15.0);
/// assert_eq!(root.len(), 2);
/// ```
pub struct Composite<T> {
    /// The value this node holds.
    pub value: T,
    children: Vec<Composite<T>>,
}

impl<T> Composite<T> {
    /// A leaf holding `value`.
    pub fn new(value: T) -> Self {
        Self {
            value,
            children: Vec::new(),
        }
    }

    /// A node holding `value` with the given children.
    pub fn with_children(value: T, children: Vec<Composite<T>>) -> Self {
        Self { value, children }
    }

    /// Appends a child.
    pub fn add(&mut self, child: Composite<T>) -> &mut Self {
        self.children.push(child);
        self
    }

    /// Removes the child at `index`, returning it, or `None` if out of range.
    pub fn remove(&mut self, index: usize) -> Option<Composite<T>> {
        if index < self.children.len() {
            Some(self.children.remove(index))
        } else {
            None
        }
    }

    /// The direct children.
    pub fn children(&self) -> &[Composite<T>] {
        &self.children
    }

    /// Depth-first, self first, and lazy.
    pub fn walk(&self) -> impl Iterator<Item = &Composite<T>> {
        let mut stack = vec![self];
        std::iter::from_fn(move || {
            let node = stack.pop()?;
            // Push in reverse so the leftmost child comes out first.
            stack.extend(node.children.iter().rev());
            Some(node)
        })
    }

    /// Adds `of(value)` over the whole subtree.
    pub fn sum(&self, of: impl Fn(&T) -> f64) -> f64 {
        self.walk().map(|node| of(&node.value)).sum()
    }

    /// How many nodes are in the subtree, including this one.
    pub fn len(&self) -> usize {
        self.walk().count()
    }

    /// Whether this node has no children.
    pub fn is_empty(&self) -> bool {
        self.children.is_empty()
    }
}

/// Flyweight — share one value per key instead of re-creating equal values.
///
/// Values come back as `Arc<V>`: they live in a map behind a lock, so they cannot
/// be lent out as `&V`. Sharing *is* the `Arc`.
///
/// ```
/// use gof_patterns::Flyweight;
/// use std::sync::Arc;
///
/// // K is annotated because a bare `&str` closure argument would infer `K = str`,
/// // which is unsized and cannot be a key.
/// let types = Flyweight::new(|name: &&str| name.to_uppercase());
/// assert!(Arc::ptr_eq(&types.get("oak"), &types.get("oak")));
/// assert_eq!(types.len(), 1);
/// ```
pub struct Flyweight<K: Eq + Hash + Clone, V, F: Fn(&K) -> V> {
    factory: F,
    cache: Mutex<HashMap<K, Arc<V>>>,
}

impl<K: Eq + Hash + Clone, V, F: Fn(&K) -> V> Flyweight<K, V, F> {
    /// A flyweight that builds values with `factory`.
    pub fn new(factory: F) -> Self {
        Self {
            factory,
            cache: Mutex::new(HashMap::new()),
        }
    }

    /// The shared value for `key`, built on first use.
    pub fn get(&self, key: K) -> Arc<V> {
        let mut cache = self.cache.lock().expect("Flyweight lock poisoned");
        Arc::clone(
            cache
                .entry(key.clone())
                .or_insert_with(|| Arc::new((self.factory)(&key))),
        )
    }

    /// How many distinct keys are held.
    pub fn len(&self) -> usize {
        self.cache.lock().expect("Flyweight lock poisoned").len()
    }

    /// Whether nothing is cached.
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Drops every cached value.
    pub fn clear(&self) {
        self.cache.lock().expect("Flyweight lock poisoned").clear();
    }
}

/// Bridge — a stable abstraction whose implementation can be swapped underneath.
///
/// Callers hold the `Bridge` and call [`Bridge::api`] per use, so a
/// [`Bridge::swap`] reaches everyone who already has it. Passing the
/// implementation into a constructor makes every holder re-wire instead.
///
/// The abstraction comes back as an `Arc`, for the same reason as everywhere else
/// here: it lives behind a lock so it can be replaced.
///
/// ```
/// use gof_patterns::Bridge;
///
/// let storage = Bridge::new(
///     |prefix: &'static str| {
///         Box::new(move |key: &str| format!("{prefix}:{key}")) as Box<dyn Fn(&str) -> String>
///     },
///     "s3",
/// );
/// assert_eq!((storage.api())("a"), "s3:a");
///
/// storage.swap("disk");
/// assert_eq!((storage.api())("a"), "disk:a");
/// ```
pub struct Bridge<Impl, Api> {
    build: Box<dyn Fn(Impl) -> Api>,
    api: RwLock<Arc<Api>>,
}

impl<Impl, Api> Bridge<Impl, Api> {
    /// Builds the abstraction over `implementation`.
    pub fn new(build: impl Fn(Impl) -> Api + 'static, implementation: Impl) -> Self {
        let build = Box::new(build);
        let api = Arc::new(build(implementation));
        Self {
            build,
            api: RwLock::new(api),
        }
    }

    /// The current abstraction. Call this per use, not once into a variable, or a
    /// later [`Bridge::swap`] will not reach you.
    pub fn api(&self) -> Arc<Api> {
        Arc::clone(&self.api.read().expect("Bridge lock poisoned"))
    }

    /// Replaces the implementation behind the abstraction.
    pub fn swap(&self, implementation: Impl) {
        *self.api.write().expect("Bridge lock poisoned") = Arc::new((self.build)(implementation));
    }
}
